timemachine/cpp/src/context.hpp

```diff
@@ -21,6 +21,7 @@ class Context {
     ~Context();
 
     void step(double lambda);
+    void finalize(double lambda);
```

Comment:
I might be missing some subtlety, but is there a reason not to just call `finalize` automatically at the end of `step` and `multiple_steps`? It seems like this is what we do in the reference implementation (https://github.com/proteneer/timemachine/blob/b548e75f226d65c502b0a53d1a53918a051bc0c9/timemachine/integrator.py#L190)

Reply:
It is called for `multiple_steps`, or really it calls the finalize on the integrator directly.

`step` is different in that we don't want to run 2 force calculation per step which is what would happen (actually three with initialization) if you called `finalize` every step. Would also require a different code path as its two half steps instead of a half step, full steps then half step. 

Basically the methods that take multiple steps should correctly finalize, but if you call step manually, to ensure correct coords/velocities you have to call finalize.

Reply:
Ah, I see. I guess I'm still a little bit unclear on the use cases for `step`, since it seems like we would prefer `multiple_steps` in most cases vs. running `step` inside of a loop`? IIUC the reference integrator does finalize after each step (https://github.com/proteneer/timemachine/blob/b548e75f226d65c502b0a53d1a53918a051bc0c9/timemachine/integrator.py#L159)

Reply:
Not sure we rely on `step` besides as a means of verification in some of our tests. I think there was also some use in the non-equilibrium stuff that is now attic'd.

The reference integrator step is different. It takes two half steps, while the C++ step is a half step, followed by a step, with finalize doing another half step.  We could have it so that on the first step of the integrator it only takes a half step, followed by full steps after that. Would make it more consistent. Will give that a shot

We could also store the last force computation like @maxentile did here: https://github.com/proteneer/timemachine/pull/783 though that doubles the memory requirement of the C++ implementation for no benefit besides .

To make `step` match for both the reference and C++ would be difficult as we would need two code paths which seems less desirable especially if `step` is not our preferred method.

Reply:
Doing a half step on the first call to `step` makes things non-reversible unless we want to add an `initialize` method as well. Which I had considered, and decided against to avoid the additional line required to make valid steps. Could easily add though.

I am comfortable with this as it seems correct and agrees with the reference, even if it is not identical in its implementation.

Reply:
Thinking about it more, there will probably always be a discontinuity between the reference and the C++. The reference is both integrator and context.

In the reference it knows to do exactly N steps, which is made up of 2 half steps and N - 1 full steps. The context on the other hand doesn't know anything about the integrator and is going to run N steps regardless of integrator, which in the case of Velocity Verlet results in N steps plus 2 half steps.

Which leads me to think agreement between the two implementations is the crucial part to verify.

@maxentile Do you have thoughts on this?

Reply:
Mostly agreed with the comments above.

* I don't know of any use cases where Python `.step` is preferred over `.multiple_steps`, besides testing.
* Making the internal implementations of `.step` match does not seem necessary -- agreement of the final state (xT, vT) of the trajectory seems the most important part to verify.
* Would adding an `initialize` method to Integrator base class help avoid the discontinuity you mention in https://github.com/proteneer/timemachine/pull/789#discussion_r942656625 ? (Context could do `initialize; step in a loop; finalize;` -- where `initialize` and `finalize` would be empty for some integrators.)

Reply:
> Would adding an initialize method to Integrator base class help avoid the discontinuity you mention in https://github.com/proteneer/timemachine/pull/789#discussion_r942656625 ? (Context could do initialize; step in a loop; finalize; -- where initialize and finalize would be empty for some integrators.)

I think so, will add an `initialize` method.

Reply:
Fixed in a1abf0ba37b3a35947907e41dba129d48c2860c5

Reply:
This solution makes sense to me. Longer term, I wonder whether the C++ integrators could own `multiple_steps` (similarly to the reference) to avoid needing to expose `initialize` and `finalize`.